File: solutions/zig/raindrops/1/raindrops.zig
// For Shree DR.MDD

const std = @import("std");

const Check = struct {
    num: u32,
    sound: []const u8,
};

const CHECKS = [_]Check{
    .{ .num = 3, .sound = "Pling" },
    .{ .num = 5, .sound = "Plang" },
    .{ .num = 7, .sound = "Plong" },
};

pub fn convert(buffer: []u8, n: u32) []const u8 {
    var stream = std.io.fixedBufferStream(buffer);

    for (&CHECKS) |*element| {
        if (n % element.num == 0) {
            stream.writer().writeAll(element.sound) catch unreachable;
        }
    }

    if (stream.pos == 0) {
        stream.writer().print("{}", .{n}) catch unreachable;
    }

    return stream.getWritten();
}

File: solutions/ocaml/raindrops/1/raindrops.ml
let prime_factor num factor word =
  if num mod factor = 0 then word else ""

let raindrop num =
  let drop = Printf.sprintf "%s%s%s"
      (prime_factor num 3 "Pling")
      (prime_factor num 5 "Plang")
      (prime_factor num 7 "Plong") in
  if drop = "" then string_of_int num else drop